src/scope.rs

```diff
@@ -87,18 +87,91 @@ impl<'scope> Scope<'scope> {
     ///
     /// Panics if not called from within a worker.
     ///
-    pub fn spawn<F>(&self, f: F)
-    where
-        F: FnOnce(&Scope<'scope>) + Send + 'scope,
-    {
+    pub fn spawn<T, S: ScopedSpawn<'scope, T>>(
+        &self,
+        scoped_work: S,
+    ) -> <S as ScopedSpawn<'scope, T>>::Output {
+        scoped_work.spawn(self)
+    }
+
+    /// Adds an additional reference to the scope's reference counter.
+    ///
+    /// Every call to this should have a matching call to
+    /// `Scope::remove_reference`, or the scope will block forever on
+    /// completion.
+    fn add_reference(&self) {
+        let counter = self.count.fetch_add(1, Ordering::SeqCst);
+        tracing::trace!("scope reference counter increased to {}", counter + 1);
+    }
+
+    /// Removes a reference from the scope's reference counter.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that there is exactly one a matching call to
+    /// `add_reference` for every call to this function, unless used within
+    /// `Scope::complete`.
+    unsafe fn remove_reference(&self) {
+        let counter = self.count.fetch_sub(1, Ordering::SeqCst);
+        tracing::trace!("scope reference counter decreased to {}", counter - 1);
+        if counter == 1 {
+            // Alerts the owning thread that the scope has completed.
+            //
+            // This should never panic, because the counter can only go to zero
+            // once, when the scope has been dropped and all work has been
+            // completed.
+            //
+            // SAFETY: The signal is passed as a reference, and is live for the
+            // duration of the function.
+            unsafe { Signal::send(&self.signal, ()) };
+        }
+    }
+}
+
+impl Drop for Scope<'_> {
+    fn drop(&mut self) {
+        // When the scope is dropped, block to prevent deallocation until the
+        // reference counter allows the scope to complete.
+        tracing::trace!("completing scope");
+        // SAFETY: This is explicitly allowed, because every scope starts off
+        // with a counter of 1. This should be the only call to
+        // `remove_reference` without a corresponding call to `add_reference`, so
+        // the only one that can cause the reference counter to drop to zero.
+        unsafe { self.remove_reference() };
+        Worker::with_current(|worker| {
+            let worker = worker.unwrap();
+            worker.wait_for_signal(&self.signal);
+        });
+    }
+}
+
+// -----------------------------------------------------------------------------
+// Generalized scoped spawn trait
+
+pub trait ScopedSpawn<'scope, T> {
```

Comment:
Nit: This `T` has a different purpose than the `T` found in its implementations. Since it's a marker-style generic, it could be switched to `M` (and same for the `T` in `Scope::spawn`).